conda/environments/cudf_dev_cuda11.5.yml

```diff
@@ -43,6 +43,7 @@ dependencies:
   - black=22.3.0
   - isort=5.6.4
   - mypy=0.782
+  - doxygen=1.8.20
```

Comment:
I installed 1.9.2 and got these warnings.

```
warning: Tag 'OUTPUT_TEXT_DIRECTION' at line 102 of file '-' has become obsolete.             
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 1089 of file '-' has become obsolete.              
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'LATEX_SOURCE_CODE' at line 1836 of file '-' has become obsolete.                
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'RTF_SOURCE_CODE' at line 1926 of file '-' has become obsolete.                  
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'DOCBOOK_PROGRAMLISTING' at line 2031 of file '-' has become obsolete.           
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
```

Since this is a `language: system` package that is not managed by pre-commit, we will need to make sure it works for slightly-older and current versions. At the time of writing:
- [Ubuntu 18.04 ships doxygen 1.8.13](https://packages.ubuntu.com/bionic/doxygen).
- [Ubuntu 20.04 ships doxygen 1.8.17](https://packages.ubuntu.com/focal/doxygen).
- [Ubuntu 22.04 ships doxygen 1.9.1](https://packages.ubuntu.com/jammy/doxygen).

I'd say that supporting 1.8.13 and newer is an ideal goal.

Reply:
I am using doxygen `1.8.20` because that's the version use to generate https://docs.rapids.ai/api/libcudf/nightly/

Reply:
pre-commit is not managing this dependency and uses a system (or conda) installation of `doxygen`, whatever is on the user's path. Users of Ubuntu 22.04 who do not use a conda environment to build would likely be fetching the system-installed apt package for doxygen 1.9.1. Those users would get warnings and the pre-commit check would fail. _Regardless_ of whether users are in a conda environment, they can make use of all other pre-commit hooks because they are _self-managed_ by pre-commit itself. Because this hook cannot be self-managed by pre-commit, we need to take care of two small things:
- Users without doxygen installed must be able to make pre-commit pass locally. In this case, CI (which has doxygen installed) will catch any issues.
- Users with newer versions of doxygen must be able to make pre-commit pass locally (because we cannot pin a dependency that is not managed by pre-commit). I think we can easily adapt our configuration to be compatible with both 1.8.x and 1.9.x, and that should solve 95% of the potential problems here.

If you'd like to see a concrete solution, I can push a small commit to help resolve both issues.

Reply:
Forgot to add -- if we solve the two problems above (no doxygen / new doxygen), it's totally fine to pin this the same way as our official documentation builds.

Reply:
I tried running `doxygen -u` to update with 1.9.2. It works, but it leaves some options that are unsupported by 1.8.x.
```
warning: ignoring unsupported tag 'RESOLVE_UNNAMED_PARAMS' at line 527, file -
warning: ignoring unsupported tag 'SHOW_HEADERFILE' at line 600, file -
warning: ignoring unsupported tag 'WARN_IF_INCOMPLETE_DOC' at line 818, file -
warning: ignoring unsupported tag 'FULL_SIDEBAR' at line 1536, file -
warning: ignoring unsupported tag 'MATHJAX_VERSION' at line 1619, file -
warning: ignoring unsupported tag 'DOT_UML_DETAILS' at line 2398, file -
warning: ignoring unsupported tag 'DOT_WRAP_THRESHOLD' at line 2407, file -
```

We'll probably need a slightly more complicated `sed` command to filter either 1.9.x warnings or 1.8.x warnings, depending on what version we pin to in the conda env.

Reply:
used `language: conda` to install doxygen=1.8.20 and use it for pre-commit hook.
I hope, this resolved the dependency issue completely.

Reply:
Up to now, I have avoided using `language: conda` in pre-commit hooks because it requires a system-installed conda. Developers can currently run pre-commit with the system Python, without using a conda environment. It is my understanding that this is a common developer use case that we should support. I would rather try to use whatever doxygen is already on the user’s path, with fallbacks / workarounds in the script as mentioned above.

Reply:
I got the requirement about system python. That's the reason for using wheel version of clang-format too!

Reply:
Exactly! We prefer pre-commit hooks that are Python (or pip-installable binaries), and `language: system` is a fallback option that may require additional work. From the list of [Supported Languages](https://pre-commit.com/#supported-languages), we really only want `python` or `system` (or possibly `script`?). All the others require the system to have other tools installed.

Reply:
Apart from `unsupported tag` warnings, there are many bogus warning in older versions `<1.8.20`. 
Adding regex to ignore all those warnings is a hacky approach, and not recommended. I will enforce the doxygen version check in `ci/check/doxygen.sh`.

Regarding updating the Doxyfile configuration for different version, unsupported tags alone can be removed to not throw  `unsupported tag` warnings on version from 1.8.20 to 1.9.1.
For doxygen versions `>1.9.1`  there are bogus warnings again. So, we cannot support them yet. (for example, default constructor without argument is throwing warning that their parameters are not all documented)

Reply:
> I will enforce the doxygen version check in ci/check/doxygen.sh.

That sounds good to me!

Reply:
PR updated. Please re-review